keep-runtime/src/main.rs

```diff
@@ -18,23 +18,152 @@
+use openssl::nid::Nid;
+use openssl::pkey::PKey;
+use openssl::rsa::Rsa;
+use openssl::x509::{X509Name, X509NameBuilder, X509};
+
+//use std::collections::HashMap;
+use std::path::Path;
+//use warp::http::StatusCode;
+use warp::Filter;
+#[derive(Serialize, Deserialize)]
+struct Payload {
+    encoding: String,
+    contents: Vec<u8>,
+}
+
 use log::info;
+/// Source of the key to use for TLS
+//pub const KEY_SOURCE: &str = "file-system";
+pub const KEY_SOURCE: &str = "generate";
 
-fn main() {
-    let _ = env_logger::try_init_from_env(env_logger::Env::default());
+#[tokio::main]
+async fn main() {
+    let args: Vec<String> = std::env::args().skip(1).collect();
+    let listen_port: u16 = args[0].parse().unwrap();
+    let (server_key, server_cert) = get_credentials_bytes();
 
-    let mut args = std::env::args().skip(1);
-    let path = args.next().unwrap();
-    let vars = std::env::vars();
+    // POST /payload
+    let workload = warp::post()
+        .and(warp::path("payload"))
+        .and(warp::body::json())
+        .and_then(payload_launch);
 
-    let bytes = std::fs::read(&path).expect("Unable to open file");
+    let routes = workload;
+    warp::serve(routes)
+        .tls()
+        .cert(&server_cert)
+        .key(&server_key)
+        //TODO - fix this so that we can bind to other IP addresses
+        .run(([127, 0, 0, 1], listen_port))
+        .await;
+}
 
-    let result = workload::run(&bytes, args, vars).expect("Failed to run workload");
+fn create_new_runtime(recvd_data: &[u8]) {
+    format!("About to attempt new runtime creation");
+    let _ = env_logger::try_init_from_env(env_logger::Env::default());
+    //TODO - get args these from main() if required
+    //    let args = std::env::args().skip(1);
+    let dummy_arr: [&str; 1] = [""];
+    let vars = std::env::vars();
 
+    let result = workload::run(recvd_data, &dummy_arr, vars).expect("Failed to run workload");
+    println!("Got result (println) {:#?}", result);
     info!("got result: {:#?}", result);
 }
+
+async fn payload_launch(payload: Payload) -> Result<impl warp::Reply, warp::Rejection> {
+    format!("Received a {} file", payload.encoding);
+    println!("Received a {} file", payload.encoding);
+    create_new_runtime(&payload.contents);
+    Ok(warp::reply::with_status(
+        "Payload received",
+        warp::http::StatusCode::OK,
+    ))
+}
+
+fn get_credentials_bytes() -> (Vec<u8>, Vec<u8>) {
+    let mut key: Vec<u8> = Vec::new();
+    let mut cert: Vec<u8> = Vec::new();
+    let (key, cert) = match KEY_SOURCE {
+        "file-system" => (get_key_bytes_fs(), get_cert_bytes_fs()),
+        "generate" => (generate_credentials()),
+        //no match!
+        _ => panic!("No match for credentials source"),
+    };
+    (key, cert)
+}
+
+//implementation for file system
+fn get_cert_bytes_fs() -> Vec<u8> {
+    let in_path = Path::new("key-material/server.crt");
+
+    let in_contents = match std::fs::read(in_path) {
```

Comment:
Maybe simpler to just propagate `std::io::Result<Vec<u8>>`?

Reply:
I'm not sure we're going to end up using this code - it's there almost as a placeholder (and not used for now), so I'm not sure it's worth spending too much time improving it.